### PR TITLE
IOS7: Add new iOSGraphicsManagers for 2D and 3D OpenGL graphics

### PR DESCRIPTION
**Pull request by larsamannen**
This commit adds the iOSGraphicsManager class which aims to replace the ios7 backend graphic handling.
It's intended to not include the iOSGraphicsManager into the ios7 backend yet. It will be included when changing the ios7 backend to derive from ModularGraphicsBackend.

### CODE REVIEW

**Comment by criezy** [comment]
At first glance the changes look good. I will try to test it tomorrow.

**Comment by larsamannen** [comment]
The PR is updated with commits adding not just a 2D graphics manager but also a 3D graphics manager.
It's quite a massive PR and might be hard to review and should therefore be tested.

There are some known "issues" that will be worked on:

- In portrait mode the game screen will be rendered in the middle of the screen which is not preferred when keyboard is shown. The game screen should move up to top of screen.
- The GameList widget in the launcher is too large when running in Debug mode (only affecting devices following Safe Area guidelines)
- The keyboard is shown in landscape mode if launching ScummVM in landscape mode.
- Shaders trying to change resolution to 640x480 will fail

**Comment by larsamannen** [comment]
@criezy I've updated the PR with some fixes and now finally resize of the frame depending on orientation and if keyboard is visible or not.
The PR is now mature enough to be merged according to me.
Please review and test when you have the possibility.

**Comment by criezy** [comment]
Great job! That's impressive.
I have not looked at the code yet, but I compiled and tested it on my iPad (iOS 12) and it works very nicely.

I have only found one issue: when rotating from landscape to portrait the display is not resized and the keyboard that appears is overlayed on part of the display. Hiding and showing the keyboard again (using the pinch gestures) solves this and the display gets resized.

Just after rotation:
image

After hiding and bringing back the keyboard everything is fine:
image

The same issue also happens in game:
image

image

**Comment by criezy** [comment]
Also when in the launcher I think it would be better not to pop up the keyboard when rotating to portrait mode since the GUI can use all the screen. When a game is running however, since it will only use part of the screen I think it still makes sense to automatically pop up the keyboard.

**Comment by criezy** [comment]
And I am adding a couple of screenshots for features that work well.

The launcher layout in fullscreen portrait is nice (although I am unsure about having the buttons at the top when using the icons):
image

image

Having access to the shaders is also great:
image

image

**Comment by larsamannen** [comment]
> Great job! That's impressive. I have not looked at the code yet, but I compiled and tested it on my iPad (iOS 12) and it works very nicely.

Thanks.

> I have only found one issue: when rotating from landscape to portrait the display is not resized and the keyboard that appears is overlayed on part of the display. Hiding and showing the keyboard again (using the pinch gestures) solves this and the display gets resized.

Hmmm, the resizing in the function adjustViewFrameForSafeArea is probably racy with the notifications received for the keyboard. I have 3 devices, one iPad mini ver 1 running iOS 9, one iPhone 12 mini and an iPad 12 Pro 12.9 both running latest version of iOS and I don't get the same problem. But I think I wrote in the commit message that I struggled a bit with the order of events. 
What I think happens in that case is when you rotate to portrait mode the keyboard will send the didShow notification before the adjustViewFrameForSafeArea is called by the system. The adjustViewFrameForSafeArea will reset the height to fullscreen if the orientation is changed.
If you could help me debug that it would be fantastic. Perhaps we should treat that as a minor issue which doesn't block this PR?

> Also when in the launcher I think it would be better not to pop up the keyboard when rotating to portrait mode since the GUI can use all the screen. When a game is running however, since it will only use part of the screen I think it still makes sense to automatically pop up the keyboard.

I can fix that

> The launcher layout in fullscreen portrait is nice (although I am unsure about having the buttons at the top when using the icons)

I agree, I think we should team up with the Android team and create a layout that is better suited for tablets and phones. But that will be another story.

**Comment by criezy** [comment]
> Perhaps we should treat that as a minor issue which doesn't block this PR?

I agree. For me this is not a blocker.

**Review by criezy — COMMENTED** [review]
Overall this looks solide. I only made a few comments.

Not automatically showing the keyboard in portrait mode when in the launcher, and fixing the timing issue about the frame resize when rotating can both be done in tree after this is merged. Also as noted in Discord we might want to add an option to force landscape mode for games.

File: ports.mk
Comment (larsamannen) [reply ports.mk:542]: Will fix

File: backends/platform/ios7/ios7_video.mm
Comment (larsamannen) [reply backends/platform/ios7/ios7_video.mm:86]: Hmmm, I don't think Apple's API doc is correct here. According to their doc layoutSublayersOfLayer is first supported in iOS10. However when looking at the doc of layoutSublayers: https://developer.apple.com/documentation/quartzcore/calayer/1410935-layoutsublayers?language=objc
supported from iOS 2.0: "The default implementation of this method calls the layoutSublayersOfLayer: method of the layer’s delegate object"
I tried to search where support was added but I found no real answer. However, I see some stackoverflow posts referring to iOS8 at least.
I've also verified on my iPad mini 1 running iOS 9.3.5 that this method is called.
I'll add a note above the function about this

**Comment by larsamannen** [comment]
Updated after comments.
I've added a commit dealing with the automatic opening of the keyboard when in game.

**Comment by larsamannen** [comment]
Pushed an update to send kInputScreenChanged on applicationResume to make sure the screen size is updated in the scenario:
 - minimize app in orientation landscape or portrait
 - resume app in a different orientation than above

**Review by criezy — APPROVED** [review]

**Pull request merged**
